src/core/dense_set.cc

```diff
@@ -197,6 +197,47 @@ bool DenseSet::Equal(DensePtr dptr, const void* ptr, uint32_t cookie) const {
   return ObjEqual(dptr.GetObject(), ptr, cookie);
 }
 
+void DenseSet::CloneBatch(unsigned len, CloneItem* items, DenseSet* other) const {
+  while (len) {
+    unsigned dest_id = 0;
+    for (unsigned i = 0; i < len; ++i) {
+      auto& src = items[i];
+      if (src.obj) {
+        // The majority of the CPU is spent in this block.
+        void* new_obj = other->ObjectClone(src.obj, src.has_ttl);
+        uint64_t hash = Hash(src.obj, 0);
+        other->AddUnique(new_obj, src.has_ttl, hash);
+        src.obj = nullptr;
+      }
+
+      const DenseLinkKey* link = src.link;
+      if (link) {
+        src.link = nullptr;
+        auto& dest = items[dest_id++];
+        DCHECK(!link->next.IsEmpty());
+        if (src.fetch_tail) {
+          // switch to the final state.
+          DCHECK(link->next.IsObject());
+          dest.obj = link->next.Raw();
+          src.fetch_tail = false;
+        } else {
+          dest.obj = link->Raw();
+          if (link->next.IsObject()) {
+            // next state - pre-terminal, fetch the final object.
+            dest.fetch_tail = true;
+            dest.link = link;
+          } else {
+            dest.link = link->next.AsLink();
+            __builtin_prefetch(dest.link, 0, 1);
+          }
+        }
+        __builtin_prefetch(dest.obj, 0, 1);
```

Comment:
I think if you move prefetch after setting dest.obj, the result should be better. prefetch needs time and there are only few instructions before this memory will be used

Reply:
I leave it to you to play and see :)
Based on my observations, this variation gave me the best possible performance and at this point the potential for further optimisations is very limited.